[JellyPenguinnn] iP

JellyPenguinnn opened the pull request:
# ChatBuddy Task Manager

> “The secret of getting ahead is getting started.” – Mark Twain [source](https://www.goodreads.com/quotes/219455-the-secret-of-getting-ahead-is-getting-started-the-secret)

**ChatBuddy** is a simple, **command-line** task manager that helps you organize your tasks effectively. With ChatBuddy, you can:

- Manage tasks like ToDos, Deadlines, and Events 📝
- Mark tasks as done ✅
- Delete tasks when completed 🔍
- Search for specific tasks 💯

## Features:
1. **Manage Tasks**:
   - ToDo
   - Deadline
   - Event
2. **Task Searching**: Quickly find tasks containing specific keywords.
3. **Task Completion**: Easily mark tasks as complete or incomplete.

### Future Enhancements:
- [x] Add support for task management.
- [ ] Implement reminder functionality (coming soon).
- [ ] Add notifications for upcoming deadlines (coming soon).

For developers, you can use the following code to launch ChatBuddy:

```java
public class ChatBuddy {
    public static void main(String[] args) {
        new ChatBuddy("data/tasks.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Level-0

Commit:
Level-1

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
Level-5&6

Commit:
latest version for week 2

Commit:
Level-7

Commit:
Merge branch-Level-7 with Level 7 features

Commit:
Implement Level 8 increment

Commit:
Merge branch-Level-8 with Level 8 increment

Commit:
Refactor code to use more OOP principles as part of A-MoreOOP increment

Commit:
Merge branch-A-MoreOOP into master

Commit:
Organized classes into packages

Commit:
Merge branch 'branch-A-Packages' into master

Commit:
A-Packages

Commit:
gradle

Commit:
A-Gradle increment

Commit:
Merge branch 'master' of https://github.com/JellyPenguinnn/ip

* 'master' of https://github.com/JellyPenguinnn/ip:
  A-Gradle increment
  A-Packages

WeiHanWong commented on java/chatbuddy/task/TaskList.java:
```diff
@@ -0,0 +1,36 @@
+        return tasks.get(index);
+    }
+
+    public int size() {
```
should be verb

NorbertLoh replied:
Perhaps
```suggestion
    public int getSize() {
```

WeiHanWong commented on java/chatbuddy/parser/Parser.java:
```diff
@@ -0,0 +1,39 @@
+        String[] parts = userInput.split(" ", 2);
+        String commandWord = parts[0];
+
+        switch (commandWord) {
```
Switch case does not follow proper indentation

WeiHanWong reviewed (commented):
LGTM. Just some nits to fix.

Commit:
Setup gradle and JUnit tests functions

Commit:
jar file created

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JavaDoc comments

Commit:
Merge branch-A-JavaDoc with JavaDoc increment

NorbertLoh commented on java/chatbuddy/command/AddCommand.java:
```diff
@@ -0,0 +1,59 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws ChatBuddyException {
+        Task task;
+        switch (type) {
+            case "T":
+                task = new ToDo(description);
+                break;
+            case "D":
+                String[] deadlineParts = description.split(" /by ");
+                if (deadlineParts.length < 2 || deadlineParts[1].trim().isEmpty()) {
+                    throw new ChatBuddyException("The due date of a deadline cannot be empty.");
+                }
+                task = new Deadline(deadlineParts[0], deadlineParts[1]);
+                break;
+            case "E":
+                String[] eventParts = description.split(" /from | /to ");
+                if (eventParts.length < 3 || eventParts[1].trim().isEmpty() || eventParts[2].trim().isEmpty()) {
+                    throw new ChatBuddyException("The start time or end time of an event cannot be empty.");
+                }
+                task = new Event(eventParts[0], eventParts[1], eventParts[2]);
+                break;
+            default:
+                throw new ChatBuddyException("Unknown task type.");
```
To follow our coding standards, the case statements should not be indented.

NorbertLoh commented on java/chatbuddy/command/DeleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    /**
+     * Executes the delete command, removing the task from the task list and updating the storage.
+     *
+     * @param tasks   The task list.
+     * @param ui      The user interface to display messages.
+     * @param storage The storage to save the updated task list.
+     * @throws ChatBuddyException If the task index is out of range.
+     */
```
I like how good your comments are.

NorbertLoh commented on java/chatbuddy/storage/Storage.java:
```diff
@@ -0,0 +1,103 @@
+     * @throws ChatBuddyException If there is an error while loading the tasks.
+     */
+    public ArrayList<Task> load() throws ChatBuddyException {
+        ArrayList<Task> tasks = new ArrayList<>();
```
I like how collections are named in plural forms.

NorbertLoh reviewed (commented):
LGTM 👍. Just some nits to fix.

Commit:
Add coding standards

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Implement find increment

Commit:
Merge branch 'branch-Level-9'

Commit:
Implement Level-10 increment

Commit:
Merge branch 'branch-Level-10'

Commit:
Task class: Add assertions to validate task creation

The Task class does not validate the task description, which can result in tasks being created with invalid or empty descriptions. Additionally, there are no checks on whether tasks can only be marked as done once, or unmarked only if already marked done.

This lack of validation could lead to invalid tasks and cause errors during task list operations.

Let's add assertions to:
- Ensure task descriptions are neither null nor empty.
- Only mark a task as done if it is not already marked done.
- Only unmark a task if it is marked as done.

Assertions help identify bugs early in the development process by catching invalid inputs and state transitions during testing. By ensuring checks are made during runtime in a development environment, we can improve code robustness without affecting production performance, as assertions can be disabled in production.

Commit:
Refactor Storage class for readability and maintainability

The Storage class currently contains long methods with multiple responsibilities, and uses magic strings for task types, which reduces readability and maintainability.

This needs to change because the code becomes harder to manage and maintain as the project grows, and magic strings lead to potential bugs and future refactoring difficulties.

Refactor the Storage class:
- Split the load() method into smaller, more focused methods (parseTask() and createStorageFile()).
- Replace magic strings with constants for task types and task statuses.
- Use try-with-resources for proper resource management of Scanner and FileWriter.

These changes improve code readability, maintainability, and reduce the chance of resource leaks, aligning with good software design principles.

JellyPenguinnn committed:
Merge pull request #1 from JellyPenguinnn/branch-A-Assertions

Task class: Add assertions to validate task creation

Commit:
Code Quality Increment

JellyPenguinnn committed:
Merge pull request #5 from JellyPenguinnn/branch-A-CodeQuality

Branch a code quality

Commit:
Add extension for update function, allows (delete, mark, unmark) multiple tasks, check for all classes, improve assertion and code quality

Commit:
Merge Extension into master

Commit:
Fix some minor problems, add user guide and Ui.png

Commit:
exception handle

Commit:
fix minor issues

Commit:
improved user guide

Commit:
fix minor issues

Commit:
Add math tutorial sheet task as todo into the task list

Commit:
Unmark the first task in the task list

Commit:
Delete the third task from the task list

Commit:
Update task 4 in the task list to CS2100 past year paper

Commit:
Mark task 3 and task 4 at the same time

Commit:
Add JavaFX runtime components for win and linus

Commit:
Update build.gradle
